src/processor/operator/var_length_extend/recursive_join.cpp

```diff
@@ -0,0 +1,119 @@
+#include "processor/operator/var_length_extend/recursive_join.h"
+
+namespace kuzu {
+namespace processor {
+
+bool ScanBFSLevel::getNextTuplesInternal(ExecutionContext* context) {
+    if (!hasExecuted) {
+        hasExecuted = true;
+        return true;
+    }
+    return false;
+}
+
+void RecursiveJoin::initLocalStateInternal(ResultSet* resultSet_, ExecutionContext* context) {
+    maxNodeOffset = nodeTable->getMaxNodeOffset(transaction);
+    srcNodeIDVector = resultSet->getValueVector(srcNodeIDVectorPos);
+    dstNodeIDVector = resultSet->getValueVector(dstNodeIDVectorPos);
+    distanceVector = resultSet->getValueVector(distanceVectorPos);
+    sspMorsel = std::make_unique<SSPMorsel>(maxNodeOffset, upperBound);
+    initLocalRecursivePlan(context);
+    bfsScanState.resetState();
+}
+
+bool RecursiveJoin::getNextTuplesInternal(ExecutionContext* context) {
+    while (true) {
+        if (bfsScanState.sizeScanned < sspMorsel->numVisitedNodes) {
```

Comment:
I think this control flow is clearer. I would still add a relatively long comment explaining the flow and the assumptions. The way the first ` if (bfsScanState.sizeScanned < sspMorsel->numVisitedNodes)` condition reads is still a bit confusing because it reads as if we output immediately as long as we visit some nodes. I'd write this above of immediately below `while(true)`.

`
There are two high level steps to RecursiveJoin. (1) BFS Computation phase: Grab a new source to do a BFS and compute an entire BFS starting from a single source; (2) Outputting phase: Once a computation finishes, we output the results in pieces of vectors to the parent operator.  These 2 steps are repeated iteratively until all sources to do a recursive computation are exhausted. The first if clause checks if we are in the outputting phase and if so, scans a vector to output and returns true. Otherwise, we compute a new BFS. 
`